mqtt/mqtt-broker/src/sys_topic.rs

```diff
@@ -0,0 +1,102 @@
+use std::collections::HashMap;
+use std::iter::FromIterator;
+
+use mqtt3::proto;
+use tracing::error;
+
+use crate::session::Session;
+use crate::ClientId;
+
+pub enum StateChange<'a> {
+    Subscriptions(&'a ClientId, Vec<&'a str>),
+    Connections(Vec<&'a ClientId>),
+    Sessions(Vec<&'a ClientId>),
+}
+
+impl<'a> StateChange<'a> {
+    pub fn get_subscriptions(client_id: &'a ClientId, session: &'a Session) -> Self {
+        Self::Subscriptions(
+            client_id,
+            session
+                .subscriptions()
+                .unwrap()
+                .keys()
+                .map(String::as_str)
+                .collect(),
+        )
+    }
+
+    pub fn clear_subscriptions(client_id: &'a ClientId) -> Self {
```

Comment:
I think you can combine this case with previous case by making session Optional.

Reply:
I changed the enum variant to hold an option. I think it makes sense to have 2 different creation functions, since they are conceptually doing different things (updating subscriptions vs. removing the message), but I combined the enum variants (before there was a clear subscriptions variant)

Reply:
Ah, I see your confusion here: 
`StateChange` factory methods have nothing to do with with actual state changes happened in the broker.
`new_subscription` method doesn't reflect that new subscription was added, it just creates a StateChange variant with corrent data. So it's better to call method `new_subscriptions_change(client_id, Session)`
1. new subscription requested: `StateChange::new_subscriptions_change(client_id, Some(session))`
2. subscription removed: `StateChange::new_subscriptions_change(client_id, Some(session))`
3. session closed: `StateChange::new_subscription_change(client_id, None)`
etc..

Reply:
what about this?

Reply:
I must not have refreshed github, b/c I didn't see this. I'll refactor the funcs